### PR TITLE
Adding async capability for external service calls. GIFs.

### PR DESCRIPTION
**Pull request by chrisdevwords**
Might need more cleanup. But have a look :mag_right:

### CODE REVIEW

**Comment by chrisdevwords** [comment]
Oh, snap! https://travis-ci.org/chrisdevwords/tacobot

File: app/imgur.js [hunk 1]
```diff
@@ -0,0 +1,157 @@
+var request = require('request');
+var $ = require('jquery-deferred');
+var util = require('./util');
+
+/**
+ * a simple module for searching Imgur.com for images
+ * or fetching the images from a public album.
+ * see https://api.imgur.com/models/image
```
Comment (learntoswim) [inline app/imgur.js]: Use proper JS Doc syntax - http://usejsdoc.org/tags-see.html

All links must be set as 

```
@see {@link https://api.imgur.com/models/image} for more information
```

File: app/imgur.js [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    this.apiKey = apiKey;
+}
+
+/**
+ * see https://api.imgur.com/models/image
+ * Searches Imgur.com for images returning a promise.
+ * The promise resolves with an array of objects
+ * If no images are found, promise resolves with an empty array.
+ * Promise rejects only on service error.
+ * @param {string} query (append "+ext:gif" for Gifs!)
```
Comment (learntoswim) [inline app/imgur.js]: Capitalize the parameters between { } - {String} / {Number}. Use proper sentence notation.

File: app/imgur.js [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+ * @param {string}_ sort (optional) defaults to top
+ * @param {number} page (optional) defaults to 0
+ * @returns {jquery-deferred}
+ */
+Imgur.prototype.getRandomFromSearch = function(q, sort, page) {
+
+    var def = $.Deferred();
+
+    this.search(q, sort, page)
+        .done(function(resp){
```
Comment (learntoswim) [inline app/imgur.js]: Follow preset coding standards of spaces for functions. 

```
.done(function (resp) { 
    //
}
```

File: app/imgur.js [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+ * @returns {jquery-deferred}
+ */
+Imgur.prototype.getAlbum = function(id) {
+
+    var _this = this;
+    var def = $.Deferred();
+    var options = {
+        uri : 'https://api.imgur.com/3/album/' + id + '/images',
+        headers : {
+            'Authorization'	: 'Client-ID ' + this.apiKey
```
Comment (learntoswim) [inline app/imgur.js]: Fix spacing between 'Authorization' and the colon

File: mock/imgurGallery.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
+module.exports = {
```
Comment (learntoswim) [inline mock/imgurGallery.js]: Move the "mock" folder to be a sub folder of "tests"

File: mock/imgurSearch.js [hunk 6]
```diff
@@ -0,0 +1,1795 @@
+module.exports = {
```
Comment (learntoswim) [inline mock/imgurSearch.js]: You can probably reduce the amount of values in this mock array to 5 max. No need to be so big.

File: tests/imgur.js [hunk 7]
```diff
@@ -0,0 +1,255 @@
+var should = require('should');
+var sinon = require('sinon');
+var request = require('request');
+var Imgur = require('../app/imgur');
+var mock = require('../mock');
+
+var albumId = 'ABEs0';
```
Comment (learntoswim) [inline tests/imgur.js]: Should be a CONSTANT -> move to a config file.

File: tests/imgur.js [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+
+});
+
+
+describe('Imgur service helper error handling', function(){
+
+    before(function(done){
+        sinon
+            .stub(request, 'get')
+            .yields(null, {statusCode:403}, mock.imgur.serviceError.apiKey);
```
Comment (learntoswim) [inline tests/imgur.js]: Use proper spacing between key/values

```
{statusCode: 403}
```

File: tests/imgur.js [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+        it('Resolves w/ an empty array if nothing is found on search', function(done){
+            var imgur = new Imgur('valid-key');
+            imgur.search('asdfkadsdflj +ext:gif nude bea arthur').always(function(resp){
+                //console.log(resp);
+                resp.should.be.a.array;
+                resp.length.should.equal(0);
+                done();
+            });
+        });
+        it('Rejects w/ an error object with the failed query if nothing is found on getRandom', function(done){
```
Comment (learntoswim) [inline tests/imgur.js]: Spacing between "it" blocks, also see /tests/responses.js for examples of proper syntax standards.

File: app/tacobot.js [hunk 10]
```diff
@@ -1,78 +1,159 @@
...
-	var index = Math.round(Math.random() * (arr.length - 1))
-	return arr[index];
+exports.roomEvent = function (data) {
+
+    var def = $.Deferred();
+
+    switch (data.event) {
+        case 'room_message':
+            def = exports.message(data);
+        break;
```
Comment (learntoswim) [inline app/tacobot.js]: Tab this in one tab.

File: app/tacobot.js [hunk 11]
```diff
@@ -1,78 +1,159 @@
...
+    var responseType = exports.getResponseType(data);
 
-	switch (data.event) {
-		case 'room_message':
-			return exports.message(data, next);
-		default:
-			return next(null, null);
-	}
+    switch(responseType) {
+        case "gif":
```
Comment (learntoswim) [inline app/tacobot.js]: Use single quotes

File: app/tacobot.js [hunk 12]
```diff
@@ -1,78 +1,159 @@
...
+ * @returns {jquery-deferred}
  */
-exports.message = function (data, next) {
+exports.imgurResponse = function (data) {
 
-	// Random response type
-	var responseType = exports.getRandomIndex(responses);
+    var def = $.Deferred();
+    var user = data.item.message.from.name;
+    var albumId = 'ABEs0';
```
Comment (learntoswim) [inline app/tacobot.js]: Should be a CONSTANT -> move to a config file.

File: app/tacobot.js [hunk 13]
```diff
@@ -1,78 +1,159 @@
...
+    var imgur = new Imgur('8ff16bbd77e6338');
```
Comment (learntoswim) [inline app/tacobot.js]: Should be a CONSTANT -> move to a config file.

**Comment by learntoswim** [comment]
- Love the addition of unit tests and the refactoring of the utilities.
- Please be mindful of the syntax standards that are set.
- Add a config file for the Imgur class

**Comment by chrisdevwords** [comment]
@karlstanton I think I got everything. :+1: 

Question: regarding spacing, punctuation, etc. Should we add .jscs file to this project. I think it requires making a decision between tabs or spaces (I'm pro spaces). 

Closest seems to be the airbnb style guilde. http://jscs.info/overview.html

Let me know and I can add one. Could add it to a precommit hook... I could also open a separate PR for this.

**Comment by learntoswim** [comment]
@chrisdevwords JSCS sounds good, I think we should include it.

**Comment by chrisdevwords** [comment]
@karlstanton re: the config, that mo

**Comment by chrisdevwords** [comment]
@karlstanton re: config - nconf seemed like a nice thing to have because it allows you manage environmental settings, like ports. Maybe for this it's overkill. I'll make a commit to take it out.

**Comment by chrisdevwords** [comment]
@karlstanton all set. :+1:

**Comment by learntoswim** [comment]
Great job, dude!

**Comment by learntoswim** [comment]
@chrisdevwords Please update the readme to show document the new commands. Also, put back in the random feature.

**Pull request merged**